crates/builder/src/graph/adj_list.rs

```diff
@@ -71,6 +71,33 @@ impl<NI: Idx, EV> AdjacencyList<NI, EV> {
             },
         };
     }
+
+    #[inline]
+    pub(crate) fn insert_mut(&mut self, source: NI, target: Target<NI, EV>) {
+        let edges = self.edges[source.index()].get_mut().unwrap();
+
+        match self.layout {
+            CsrLayout::Sorted => match edges.binary_search(&target) {
+                Ok(i) => edges.insert(i, target),
+                Err(i) => edges.insert(i, target),
+            },
+            CsrLayout::Unsorted => edges.push(target),
+            CsrLayout::Deduplicated => match edges.binary_search(&target) {
+                Ok(_) => {}
+                Err(i) => edges.insert(i, target),
+            },
+        };
```

Comment:
@knutwalker this is duplicated in the method above. I was wondering if we can write this generic to work on `&mut Vec<Target<NI, EV>>` and `RwLockWriteGuard<Vec<Target<NI, EV>>>`

Reply:
@s1ck `RwLockWriteGuard<Vec<Target<NI, EV>>>` should deref to a `&mut Vec<Target<NI, EV>>` so can implement one in terms of the other.

Reply:
@s1ck `RwLockWriteGuard<Vec<Target<NI, EV>>>` should deref to a `&mut Vec<Target<NI, EV>>` so can implement one in terms of the other.

Reply:
@s1ck `RwLockWriteGuard<Vec<Target<NI, EV>>>` should deref_mut to a `&mut Vec<Target<NI, EV>>` so can implement one in terms of the other.

Reply:
right, I tried that but got mislead by another borrow problem. works now! thx